### PR TITLE
Integrating HeuristicCompletion

### PR DESCRIPTION
**Pull request by tesonep**
Code completion engine using generators. Instead of query the entire system and generate lots of potentially intermediate collections that we need to concatenate, copy, grow, iterate all those sub-collections in order. Generators provide stream-like access to those collections (and sequences of collections) without scanning the full system eagerly.

Original Implementation available at [https://github.com/guillep/complishon](https://github.com/guillep/complishon)
Thanks to @guillep

### CODE REVIEW

**Comment by pavel-krivanek** [comment]
it is not directly related to this package, but we should move the class `CompletionEngine` to `NECompletion` package to avoid dependency on NECompletion-Morphic

**Comment by pavel-krivanek** [comment]
Tests should be in a separate package

**Comment by pavel-krivanek** [comment]
please, check Critic Browser results

File: src/HeuristicCompletion/CoLiteralMessageHeuristic.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoLiteralMessageHeuristic.class.st]:
```suggestion
I implement a heuristic for messages sent to literals.
```

File: src/HeuristicCompletion/CoLiteralMessageHeuristic.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoLiteralMessageHeuristic.class.st]:
```suggestion
If there is a literal as a receiver, create a fetcher for messages of its class.
```

File: src/HeuristicCompletion/CoLocalScopeVariablesHeuristic.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoLocalScopeVariablesHeuristic.class.st]:
```suggestion
I implement a heuristic for local variables accessible from the current AST node.
```

File: src/HeuristicCompletion/CoClassBasedFetcher.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoClassBasedFetcher.class.st]:
```suggestion
- I define #forHierarchy that decorates myself with a HierarchyComplishonFetcher that does a search for the entire hierarchy.
```

File: src/HeuristicCompletion/CoBenchmarkPackage.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoBenchmarkPackage.class.st]:
```suggestion
I mainly give access to methods in the package.
```

File: src/HeuristicCompletion/CoCompletionEngine.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoCompletionEngine.class.st]:
```suggestion
See my superclass for details.
```

File: src/HeuristicCompletion/CoGlobalVariablesHeuristic.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoGlobalVariablesHeuristic.class.st]:
```suggestion
I implement a heuristic for global variables accessible in the completion environment.
```

File: src/HeuristicCompletion/CoSelfMessageHeuristic.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoSelfMessageHeuristic.class.st]:
```suggestion
I implement a heuristic for self message sends.
```

File: src/HeuristicCompletion/CoSuperMessageHeuristic.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoSuperMessageHeuristic.class.st]:
```suggestion
I implement a heuristic for self message sends.
```

File: src/HeuristicCompletion/CoHeuristicBuilder.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoHeuristicBuilder.class.st]:
```suggestion
I am a heuristic builder.
```

File: src/HeuristicCompletion/CoHeuristicBuilder.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoHeuristicBuilder.class.st]:
```suggestion
I am composed of several small heuristics and generate a chain of responsibility from them in the right order.
```

File: src/HeuristicCompletion/CoHeuristicBuilder.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoHeuristicBuilder.class.st]:
```suggestion
I am also used as a hook, that can be redefined in subclasses and build heuristics differently.
```

File: src/HeuristicCompletion/CoGlobalVariableMessageHeuristic.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoGlobalVariableMessageHeuristic.class.st]:
```suggestion
 - if it is a class complete messages in its class side
```

File: src/HeuristicCompletion/CoGlobalVariableMessageHeuristic.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoGlobalVariableMessageHeuristic.class.st]:
```suggestion
I implement a heuristic for messages sent to globals.
```

File: src/HeuristicCompletion/CoGlobalVariableMessageHeuristic.class.st
Comment (pavel-krivanek) [reply src/HeuristicCompletion/CoGlobalVariableMessageHeuristic.class.st]: class complete?

File: src/HeuristicCompletion/CoSystemEnvironment.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoSystemEnvironment.class.st]:
```suggestion
I know a system dictionary and can be initialized with, for example, Smalltalk globals.
```

File: src/HeuristicCompletion/CoASTHeuristicsResultSetBuilder.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoASTHeuristicsResultSetBuilder.class.st]:
```suggestion
I am an AST based heuristics builder.
```

File: src/HeuristicCompletion/CoResultSetBuilder.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoResultSetBuilder.class.st]:
```suggestion
I am an AST based heuristics builder.
```

File: src/HeuristicCompletion/CoResultSetBuilder.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoResultSetBuilder.class.st]:
```suggestion
The completion engine configures me with a completion context and eventually sends me the #buildCompletion message.
```

File: src/HeuristicCompletion/CoFetcher.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoFetcher.class.st]:
```suggestion
This way any, use of #next will be done step by step. 
```

File: src/HeuristicCompletion/CoAvoidRepeatedFetcher.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoAvoidRepeatedFetcher.class.st]:
```suggestion
I am mainly used in the context of code completion of messages, although I can be used in another context.
```

File: src/HeuristicCompletion/CoAvoidRepeatedFetcher.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoAvoidRepeatedFetcher.class.st]:
```suggestion
For example, when looking to autocomplete messages for ==self==, a heuristic may choose to show the selectors in the class, then the selectors in the superclass, and so on, following the inheritance chain. However, in the class hierarchy, we may observe redefined methods, so a naive fetcher will first show #redefined from the subclass and then #redefined again from the superclass.
```

File: src/HeuristicCompletion/CoStaticBenchmarks.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoStaticBenchmarks.class.st]:
```suggestion
I take all existing call sites in a scope (a class, a package) and run the completion with different prefixes (2 to 8), asking for 10 results.
```

File: src/HeuristicCompletion/CoStaticBenchmarks.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoStaticBenchmarks.class.st]:
```suggestion
Then I check if the selector at that call site is in those elements, and in what position.
```

File: src/HeuristicCompletion/CoASTNodeFetcherHeuristic.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoASTNodeFetcherHeuristic.class.st]:
```suggestion
I am implemented as a chain of responsibility: I treat the build request if possible, otherwise, I forward the request to the next heuristic.
```

File: src/HeuristicCompletion/CoStatisticsCompletionEntry.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoStatisticsCompletionEntry.class.st]:
```suggestion
See CoStatisticsHeuristicWrapper for more details.
```

File: src/HeuristicCompletion/CoWorkspaceVariablesHeuristic.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoWorkspaceVariablesHeuristic.class.st]:
```suggestion
I implement a heuristic for variables in a workspace/playground.
```

File: src/HeuristicCompletion/CoStatistics.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoStatistics.class.st]:
```suggestion
Accepted events also contain the text accepted, the index, and the heuristic.
```

File: src/HeuristicCompletion/CoNullFetcherHeuristic.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoNullFetcherHeuristic.class.st]:
```suggestion
I am a null fetcher heuristic, returning an empty fetcher.
```

File: src/HeuristicCompletion/CoSessionMessagesHeuristic.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoSessionMessagesHeuristic.class.st]:
```suggestion
I am a heuristic that autocompletes messages based on a session vocabulary.
```

File: src/HeuristicCompletion/CoASTHeuristicsResultSetBuilder.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoASTHeuristicsResultSetBuilder.class.st]:
```suggestion
2) I extract the currently typed text and set is as an initial filter
```

File: src/HeuristicCompletion/CoInitializeInferencedMessageHeuristic.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoInitializeInferencedMessageHeuristic.class.st]:
```suggestion
I am a heuristic using depth-bound type inference to type variables.
```

File: src/HeuristicCompletion/CoMethodVariableFetcher.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoMethodVariableFetcher.class.st]:
```suggestion
From a given AST node, I walk up the tree looking for scopes and yield the variables I find.
```

File: src/HeuristicCompletion/CoStatisticsHeuristicWrapper.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoStatisticsHeuristicWrapper.class.st]:
```suggestion
I am an heuristic decorator that wraps a heuristic.
```

File: src/HeuristicCompletion/CoStatisticsHeuristicWrapper.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoStatisticsHeuristicWrapper.class.st]:
```suggestion
When asked to build a fetcher, I build the fetcher of the wrapped heuristic and decorate it with a mapping fetcher. This mapping fetcher wraps the completion entries returned by the fetcher with a logging entry.
```

**Comment by tesonep** [comment]
Thanks @pavel-krivanek I will check it tomorrow!

File: src/HeuristicCompletion/CoEmptyFetcher.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoEmptyFetcher.class.st]:
```suggestion
By default, I do nothing. I'm the NullObject of the fetcher hierarchy.
```

File: src/HeuristicCompletion/CoEntryAcceptedEvent.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoEntryAcceptedEvent.class.st]:
```suggestion
I represent the event of having accepted entry in the candidate list of the code completion.
```

File: src/HeuristicCompletion/CoInstanceAccessibleVariablesHeuristic.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoInstanceAccessibleVariablesHeuristic.class.st]:
```suggestion
I implement a heuristic for variables accessible from the instance side, such as instance variables and class variables.
```

File: src/HeuristicCompletion/CoStatisticsHeuristicBuilder.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoStatisticsHeuristicBuilder.class.st]:
```suggestion
I am a heuristic builder that builds statistic-enabled fetchers.
```

File: src/HeuristicCompletion/CoStatisticsHeuristicBuilder.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoStatisticsHeuristicBuilder.class.st]:
```suggestion
I am also used as a hook, that can be redefined in subclasses and build heuristics differently.
```

File: src/HeuristicCompletion/CoVariableValueMessageHeuristic.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoVariableValueMessageHeuristic.class.st]:
```suggestion
I implement a heuristic for messages sent to variables whose value is available in the context, for example in a debugger, workspace or inspector.
```

File: src/HeuristicCompletion/CoStatisticsASTHeuristicsResultSetBuilder.class.st
Comment (pavel-krivanek) [inline src/HeuristicCompletion/CoStatisticsASTHeuristicsResultSetBuilder.class.st]:
```suggestion
I am a result set builder that builds heuristics using statistic support.
```

**Comment by pavel-krivanek** [comment]
when you try to open the code preview on the selected completion (cmd->), the error: `Instance of CoUnknownMessageHeuristic did not understand #receiver` raises

**Comment by pavel-krivanek** [comment]
btw., the word "complishon" used on many places (and for original name of the package) is very strange, does it have some reasonable technical background?

**Comment by pavel-krivanek** [comment]
Some short package comment in Manifest would be great

**Comment by pavel-krivanek** [comment]
in `SystemNavigation` extensions like `SystemNavigation>>#allSentMessagesInClass:do:`, the `self environment` should be used instead of `Smalltalk`

**Comment by pavel-krivanek** [comment]
place `CompletionEngine` extensions directly into this class. It would make dependency of this new package on UI (NECompletion has it already so we should not create new one to solve)

**Comment by guillep** [comment]
> btw., the word "complishon" used on many places (and for original name of the package) is very strange, does it have some reasonable technical background?

At the beginning this started as a prototype, and Complishon is just a joke: "completion" badly written. It should be banned. Pablo passed around and remove lots of those ^^

**Comment by guillep** [comment]
> when you try to open the code preview on the selected completion (cmd->), the error: `Instance of CoUnknownMessageHeuristic did not understand #receiver` raises

Ah, I could never make cmd-> work in my mac, I think a problem with the events, so I could never test it :/

**Comment by guillep** [comment]
> Some short package comment in Manifest would be great

We could add something from the original readme.
https://github.com/guillep/complishon/

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/guillep/complishon">guillep/complishon</a></strong></div><div>Contribute to guillep/complishon development by creating an account on GitHub.</div></blockquote>

**Comment by guillep** [comment]
> place `CompletionEngine` extensions directly into this class. It would make dependency of this new package on UI (NECompletion has it already so we should not create new one to solve)

I think the right move would be to split HeuristicCompletion in HeuristicCompletion-UI and HeuristicCompletion-Model or Core or something.

**Comment by guillep** [comment]
@tesonep if this can wait until next week, I can work on it on monday.

**Comment by guillep** [comment]
> > when you try to open the code preview on the selected completion (cmd->), the error: `Instance of CoUnknownMessageHeuristic did not understand #receiver` raises
> 
> Ah, I could never make cmd-> work in my mac, I think a problem with the events, so I could never test it :/

@pavel-krivanek, could you share the stack? I'll try to make the detail popup in osx, because I think it never worked, but I'll see...

**Comment by guillep** [comment]
> > > when you try to open the code preview on the selected completion (cmd->), the error: `Instance of CoUnknownMessageHeuristic did not understand #receiver` raises
> > 
> > 
> > Ah, I could never make cmd-> work in my mac, I think a problem with the events, so I could never test it :/
> 
> @pavel-krivanek, could you share the stack? I'll try to make the detail popup in osx, because I think it never worked, but I'll see...

Well, it turns out that the code is not so well written, so Cmd+Shift+right and Cmd+Ctrl+Shift+right all activate the "shortcut". Maybe it was on purpose, but that allows me to test without changing all the events :)

**Comment by guillep** [comment]
Hmm I see somehow the classes lost their class original comments (see my original repo has them https://github.com/guillep/complishon/)

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/guillep/complishon">guillep/complishon</a></strong></div><div>Contribute to guillep/complishon development by creating an account on GitHub.</div></blockquote>

**Comment by guillep** [comment]
Ok, this should be almost ready. @pavel-krivanek maybe you want to give a final review.
Let's also wait for the CI.

**Comment by pavel-krivanek** [comment]
I tested: https://ci.inria.fr/pharo-ci-jenkins2/job/Test%20pending%20pull%20request%20and%20branch%20Pipeline/job/PR-5996/11/artifact/bootstrap-cache/Pharo9.0-PR-64bit-d4cf495.zip

the return issue is still there. The image with the open debugger is here: http://mysharegadget.com/914783218

<blockquote></blockquote>
<blockquote><div><strong><a href="http://mysharegadget.com/914783218">review.zip review.image, review.changes, PharoDebug.log, stderr, stdout, pharo.version, Pharo9.0-64bit-d4cf495.sources,  - ShareGadget.com Share Send files by air mail - Share your files - Upload your file here</a></strong></div></blockquote>

**Comment by guillep** [comment]
That is a spotter issue, isn't it? Not an autocompletion issue.
That should be fixed in #5992 , not in here. @tesonep can you confirm?

**Comment by pavel-krivanek** [comment]
ah, right, sorry. I mixed the issues

**Review by pavel-krivanek — APPROVED** [review]
issue fixed

**Comment by guillep** [comment]
- there is a release test failing related to manifest naming: testManifestNamesAccordingToPackageNames (which is strange because the manifest was automatically created by calypso...)
- there is a non-related fleaky test failure (logged in https://github.com/pharo-project/pharo/issues/6052)

**Comment by guillep** [comment]
Tests fixed

**Comment by tesonep** [comment]
I think we can integrate it, @guillep @pavel-krivanek ?

**Comment by tesonep** [comment]
I am rerunning because there have been a crash, but I think it will be ready.

**Comment by Ducasse** [comment]
Nice :)

**Pull request merged**
